employees/views.py

```diff
@@ -37,41 +39,80 @@ def query_as_dict(query_set):
 class ReportList(APIView):
     renderer_classes = [renderers.TemplateHTMLRenderer]
     template_name = 'employees/report_list.html'
+    reports_dict = {}
+    project_form = ''
     permission_classes = (
         permissions.IsAuthenticated,
     )
 
     def get_queryset(self):
         return Report.objects.filter(author=self.request.user).order_by('-date', 'project__name')
 
+    def _add_project(self, serializer, project):
+        project.members.add(self.request.user)
+        project.full_clean()
+        project.save()
+        project.refresh_from_db()
+        serializer.fields['project'].initial = project
+
+    def _create_serializer(self):
+        reports_serializer = ReportSerializer(context={'request': self.request}, )
+        reports_serializer.fields['project'].queryset = \
+            Project.objects.filter(
+                members__id=self.request.user.id
+            ).order_by('name')
+        return reports_serializer
+
+    def initial(self, request, *args, **kwargs):
+        super().initial(request, *args, **kwargs)
+        self.reports_dict = query_as_dict(self.get_queryset())
+        self.project_form = ProjectJoinForm(
+            queryset=Project.objects.exclude(members__id=self.request.user.id).order_by('name'),
+        )
+
     def get(self, request):
-        reports_serializer = ReportSerializer(context={'request': request})
-        reports_dict = query_as_dict(self.get_queryset())
         return Response({
-            'serializer': reports_serializer,
-            'reports_dict': reports_dict,
+            'serializer': self._create_serializer(),
+            'reports_dict': self.reports_dict,
             'UI_text': ReportListStrings,
+            'project_form': self.project_form,
         })
 
     def post(self, request):
         reports_serializer = ReportSerializer(data=request.data, context={'request': request})
-        reports_dict = query_as_dict(self.get_queryset())
+
+        if 'join' in request.POST:
+            project_id = request.POST['projects']
+            project = Project.objects.get(id=int(project_id))
+            self._add_project(serializer=reports_serializer, project=project)
+            self.project_form = ProjectJoinForm(
+                queryset=Project.objects.exclude(members__id=self.request.user.id).order_by('name'),
+            )
+            reports_serializer = self._create_serializer()
+            reports_serializer.fields['project'].initial = project
+            return Response({
+                'serializer': reports_serializer,
+                'reports_dict': self.reports_dict,
+                'UI_text': ReportListStrings,
+                'project_form': self.project_form,
+            })
+
         if not reports_serializer.is_valid():
```

Comment:
Also please change it into `elif`

Reply:
But how? It's nested inside the previous if statement and so is reports_serializer initialisation and other variables required further. It is not an alternative statement in any way, it is purposefully completely dependent on the previous one.

Reply:
What are the benefits of it?